File: Leetcode/_234_Palindrome_Linked_List.java
// link - https://leetcode.com/problems/palindrome-linked-list/

import java.util.ArrayList;

public class _234_Palindrome_Linked_List {
    // Store it as an ArrayList approach
    public boolean isPalindromeNaive(ListNode head) {
        ArrayList<Integer> list = new ArrayList<>();

        while (head != null) {
            list.add(head.val);
            head = head.next;
        }

        for (int i = 0, j = list.size() - 1; i < j; i++, j--) {
            if (list.get(i) != list.get(j))
                return false;
        }

        return true;
    }

    // Reverse the right half appraoch
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode revHead = reverseList(slow.next);

        while (head != null && revHead != null) {
            if (head.val != revHead.val) {
                return false;
            }
            head = head.next;
            revHead = revHead.next;
        }

        return true;
    }

    public ListNode reverseList(ListNode start) {
        ListNode prev = null;
        while (start != null) {
            ListNode next = start.next;
            start.next = prev;
            prev = start;
            start = next;
        }
        return prev;
    }
}

File: Leetcode/python/_234_Palindrome_Linked_List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional


class Solution:
    # Fast and Slow Pointers - Time = O(n) - Space = O(1)
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # Get to the middle
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        start = head
        # Reverse list from middle to the end
        revStart = self.rev(slow)

        # Now check if start and revStart match
        while start and revStart:
            if start.val != revStart.val:
                return False
            start = start.next
            revStart = revStart.next

        return True

    def rev(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        while head:
            temp = head.next
            head.next = prev
            prev = head
            head = temp
        return prev
